### PR TITLE
add PN9 Whitener

### PR DESCRIPTION
**Pull request by ffierling**
**PN9 Whitener**
- XORs each char in PDU of input message against the PN9 pseudo-random bit sequence described in section 17.2.3 of the 802.15.4-2015 specification.

Tested with:
- Microchip AT86RF215
- Analog Devices ADF7023

### CODE REVIEW

File: lib/pn9_whitener_impl.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    class pn9_whitener_impl : public pn9_whitener
+    {
+     private:
+      const static int MAX_PREAMBLE_LEN = 8;
+      const static int MAX_SFD_LEN = 4;
+      const static int MAX_PHR_LEN = 8;
+      const static int MAX_PSDU_LEN = 2048; // Includes CRC, FCS, or MFR
+      const static int MAX_PPDU_LEN = MAX_PREAMBLE_LEN + MAX_SFD_LEN + MAX_PHR_LEN + MAX_PSDU_LEN;
+      const static int MAX_FCS_LEN = 4; // Same as MFR
+
```
Comment (bastibl) [inline lib/pn9_whitener_impl.h]: not used

File: lib/pn9_whitener_impl.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      // Length in bytes over which a 511-bit PN9 sequence repeats
+      const static int PN9_LEN = (1<<9)-1;
+
+      unsigned char* d_pn9;
+
```
Comment (bastibl) [inline lib/pn9_whitener_impl.h]: that should be statically allocated

File: lib/pn9_whitener_impl.cc [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    /*
+     * The private constructor
+     */
+    pn9_whitener_impl::pn9_whitener_impl(uint16_t seed)
+      : gr::block("pn9_whitener",
+              gr::io_signature::make(0,0,0),
+              gr::io_signature::make(0,0,0))
+    {
+      uint16_t pn = seed;
```
Comment (bastibl) [inline lib/pn9_whitener_impl.cc]: shouldn't the seed be different for each frame?

File: include/ieee802_15_4/pn9_whitener.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    class IEEE802_15_4_API pn9_whitener : virtual public gr::block
+    {
+     public:
+      typedef boost::shared_ptr<pn9_whitener> sptr;
+
+      /*!
+       * \param seed Configurable starting seed. Defaults to all ones.
+       */
+      static sptr make(uint16_t seed = ~0); // SUN PHY default
```
Comment (bastibl) [inline include/ieee802_15_4/pn9_whitener.h]: maybe better 0xffff?

File: include/ieee802_15_4/CMakeLists.txt [hunk 5]
```diff
@@ -40,6 +40,6 @@ install(FILES
     rime_stack.h
     zeropadding_b.h
     zeropadding_removal_b.h
-    DESTINATION ${GR_INCLUDE_DIR}/gnuradio/ieee802_15_4
+    pn9_whitener.h DESTINATION ${GR_INCLUDE_DIR}/gnuradio/ieee802_15_4
```
Comment (bastibl) [inline include/ieee802_15_4/CMakeLists.txt]: sorting

File: lib/CMakeLists.txt [hunk 6]
```diff
@@ -49,6 +49,7 @@ list(APPEND ieee802_15_4_sources
     uc_connection.cc
     zeropadding_b_impl.cc
     zeropadding_removal_b_impl.cc
+    pn9_whitener_impl.cc
```
Comment (bastibl) [inline lib/CMakeLists.txt]: sorting

**Comment by ffierling** [comment]
Okay, stay tuned then.

**Pull request closed without merging**